### ISSUE TITLE
EF Core 5.0.1 problem with enum store in tinyint

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/efcore#23858]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/EF-Core-501-problem-with-enum-store-in/1285586)._

---
<p>We store enumerations in the database as tinyint to save space:</p><p><a href="https://developercommunity.visualstudio.com/storage/attachments/166205-tbltodo.png"> Image:166205-tbltodo.png</a></p><p>Enum Definition:</p><pre>   Enum TdoOwner
      &lt;_DisplayText(&quot;Global&quot;)&gt; [Global] = 1
      &lt;_DisplayText(&quot;Address&quot;)&gt; Address = 2
      &lt;_DisplayText(&quot;Contact&quot;)&gt; Contact = 3
      &lt;_DisplayText(&quot;Abonnement&quot;)&gt; Abonnement = 4
      &lt;_DisplayText(&quot;Order&quot;)&gt; Order = 5
   End Enum</pre><p>Entity Definition:</p><pre>Public Overridable Property TdoOwnerTyp As Byte?
</pre><p>ModelBuilder definition:</p><pre>modelBuilder.Entity(Of tblToDo)(). Property(Of Byte?) (Function(x) x.TdoOwnerTyp). HasColumnName(&quot;TdoOwnerTyp&quot;). ValueGeneratedNever(). HasPrecision(3, 3). HasDefaultValueSql(&quot;0&quot;)
</pre><p>If we create a query like</p><pre>From t In objStore.tblToDo
Where t.TdoOwnerTyp = TdoOwner.Address
Order By t.TdoCloseDate Descending
Select New With {t.TdoKey, t.TdoTargetDate, t.TdoOwnerTyp, t.TdoTargetKey}
</pre><p>EF Core creates an SQL Statement which starts with</p><pre>DECLARE @__$VB$Local_intOwnerType_0 tinyint = CAST(Address AS tinyint)
</pre><p>Why does EF Core translate an enum by Name instead of Value?</p> <p>We tried to define the enum as Byte. Does not help. <br>We also tried to define the entity as Integer? instead of Byte?. Does not help.</p> <p>Only extending the model builder by &quot;HasConversion(Of Integer)()&quot; helped to let create EF Core the the SQL statement as:</p><pre>DECLARE @__$VB$Local_intOwnerType_0 tinyint = CAST(2 AS tinyint)
</pre><p>But as soon as we want to create a list of the executed query we get a System.InvalidCastException (Unable to cast object of type 'System.Byte' to type 'System.Int32').</p> <p>What is the correct way to let EF Core to handle Enum values stored as tinyint=</p><p>Thank you for your feedback</p>

---
#### Feedback Bot on 12/13/2020, 11:16 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by ajcvickers** [comment]
Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by travico** [comment]
Et voilà. Please find attached a sample project.
On start it will fill up a small in-memory database and trigger a query having an enum as byte? (tinyint) in the where clause.

The mistake in the generated sql statement is using the enum name (Contact) instead of enum value (3):

SELECT t.TdoKey, t.TdoTargetDate, t.TdoOwnerTyp, t.TdoTargetKey
FROM tblToDo AS t
WHERE COALESCE((t.TdoOwnerTyp = **_Contact_**) AND t.TdoOwnerTyp IS NOT NULL, 0)
ORDER BY t.TdoTargetDate DESC

Thank you for investigating.

[WinFormsApp2.zip](https://github.com/dotnet/efcore/files/5808401/WinFormsApp2.zip)

**Comment by ajcvickers** [comment]
Note for triage: query expression tree:

```
DbSet<tblToDo>()
    .Where(t => ConvertChecked(t.TdoOwnerTyp, Nullable`1)
     == ConvertChecked(ConvertChecked(Contact, Int32), Nullable`1)
     ?? False)
    .OrderByDescending(t => t.TdoTargetDate)
    .Select(t => new VB$AnonymousType_0<int, Nullable<DateTime>, Nullable<byte>, Nullable<int>>(
        TdoKey = t.TdoKey, 
        TdoTargetDate = t.TdoTargetDate, 
        TdoOwnerTyp = t.TdoOwnerTyp, 
        TdoTargetKey = t.TdoTargetKey
    ))
```

**Comment by smitpatel** [comment]
Re-sending this back to triage. This is not a regression from 3.1. The same bug existed in 3.1. Double converts are generated by VB only, C# doesn't generate them. Making this query work is not minor bug fix, it is more detailed enhancement which can be risky.

**Comment by travico** [comment]
Just tried this issue with VB .NET SDK 6.0.100-preview.7.21379.14 and EF Core 6.0.0-preview.7.21378.4
Still the same problem.
Is there any hope to get this solved in EF Core 6?

**Comment by ajcvickers** [comment]
@travico Yes, this is still scheduled to be fixed in 6.0. However, it is a relatively low priority bug--no votes, for example--so there is some chance it will be punted if we run out for time.

**Comment by travico** [comment]
Hello Arthur,
Thank you for your feedback and no problem.
I solved it with a custom expression visitor. 
Let's hope there are no dramatic changes in expression visits until final release.
